apiserver/common/registry.go

```diff
@@ -49,30 +49,50 @@ func RegisterFacadeForFeature(name string, version int, factory facade.Factory,
 
 // validateNewFacade ensures that the facade factory we have has the right
 // input and output parameters for being used as a NewFoo function.
-func validateNewFacade(funcValue reflect.Value) error {
+func validateNewFacade(funcValue reflect.Value) (bool, error) {
 	if !funcValue.IsValid() {
-		return fmt.Errorf("cannot wrap nil")
+		return false, fmt.Errorf("cannot wrap nil")
 	}
 	if funcValue.Kind() != reflect.Func {
-		return fmt.Errorf("wrong type %q is not a function", funcValue.Kind())
+		return false, fmt.Errorf("wrong type %q is not a function", funcValue.Kind())
 	}
 	funcType := funcValue.Type()
 	funcName := runtime.FuncForPC(funcValue.Pointer()).Name()
-	if funcType.NumIn() != 3 || funcType.NumOut() != 2 {
-		return fmt.Errorf("function %q does not take 3 parameters and return 2",
-			funcName)
-	}
 
-	type nastyFactory func(
-		st *state.State,
-		resources facade.Resources,
-		authorizer facade.Authorizer,
-	) (
-		interface{}, error,
+	badSigError := errors.Errorf(""+
+		"function %q does not have the signature func "+
+		"(facace.Context) (*Type, error), or "+
+		"(*state.State, facade.Resources, facade.Authorizer) (*Type, error)", funcName)
+
+	if funcType.NumOut() != 2 {
+		return false, errors.Trace(badSigError)
+	}
+	var (
+		facadeType reflect.Type
+		nice       bool
 	)
-	facadeType := reflect.TypeOf((*nastyFactory)(nil)).Elem()
+	inArgCount := funcType.NumIn()
+
+	switch inArgCount {
+	case 1:
+		type niceFactory func(facade.Context) (interface{}, error)
+		facadeType = reflect.TypeOf((*niceFactory)(nil)).Elem()
+		nice = true
+	case 3:
+		type nastyFactory func(
```

Comment:
Maybe it would be a bit clearer if nice/nastyFactory were defined outside the switch?